ArduPlane/parachute.cpp

```diff
@@ -56,3 +61,52 @@ bool Plane::parachute_manual_release()
 
     return true;
 }
+
+/*
+  parachute_emergency_check - trigger the release of the parachute
+  automatically if a critical situation is detected
+*/
+void Plane::parachute_emergency_check()
+{
+    uint32_t now = millis();
+
+    // exit immediately if parachute or automatic release is not enabled, or already released
+    if (!parachute.auto_enabled() || parachute.released()) {
+        parachute.control_loss_ms(0);
+        return;
+    }
+
+    // only automatically release in AUTO mode
+    if (control_mode != AUTO) {
+        parachute.control_loss_ms(0);
+        return;
+    }
+
+    // do not release if taking off or landing
+    if (auto_state.takeoff_complete == false || mission.get_current_nav_cmd().id == MAV_CMD_NAV_LAND) {
+        parachute.control_loss_ms(0);
+        return;
+    }
+
+    if (relative_altitude() < parachute.alt_min()) {
+        parachute.control_loss_ms(0);
+        return;
+    }
+
+    // do not release if we are flying within given error
+    if (altitude_error_cm < parachute.auto_error_cm()) {
+        parachute.control_loss_ms(0);
+        return;
+    }
+
+    // at this point we consider control lost
+    if (parachute.control_loss_ms() == 0) {
+        gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Crash: Starting to lose control, %d m error", altitude_error_cm * 0.01f);
+        parachute.control_loss_ms(now);
+    } else if (now - parachute.control_loss_ms() > PARACHUTE_CHECK_TRIGGER_MS) {
+        // we have lost control for too long
+        gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Crash: Critical altitude error %d m", altitude_error_cm * 0.01f);
+        parachute_release();
+        parachute.control_loss_ms(0);
```

Comment:
I think that `parachute.released()` check should kick in after `parachute_release()` (line 74).